### PR TITLE
docs: add a UDPIngress guide

### PR DESCRIPTION
**Pull request by shaneutt**
### Review
@Bradamant3 

### Summary

This is another iteration on the KIC 2.x docs which are in flight and adds a `UDPIngress` guide page.

### Reason

Resolves https://github.com/Kong/docs.konghq.com/issues/2903

### Testing

I tested the guide manually running through each step to ensure that everything up to and including the verification step worked.

Everything worked minus a bug a found in the chart which is now reported as:

- https://github.com/Kong/charts/issues/445

And will be fixed before 2.0.0 GA.

### CODE REVIEW

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+
+This example assumes you have the `dig` command available on your local system, but if
+you don't look to your operating system documentation for a similar DNS lookup tool.
+
+Firstly you'll need to retrieve the IP address of the UDP load balancer service that we
+configured in previous sections:
+
+```shell
+$ export KONG_UDP_ENDPOINT="$(kubectl -n <NAMESPACE> get service <RELEASE_NAME>-kong-udp-proxy \
+    -o=go-template='{{range .status.loadBalancer.ingress}}{{.ip}}{{end}}')"
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:299]:
```suggestion
$ export KONG_UDP_ENDPOINT="$(kubectl -n {NAMESPACE} get service {RELEASE_NAME}-kong-udp-proxy \
    -o=go-template='{% raw %}{{range .status.loadBalancer.ingress}}{{.ip}}{{end}}{% endraw %}')"
```
In Jekyll, `{{ }}` refer to variables, so it's trying to parse the tag and not getting anything. These `{% raw %}` tags tell Jekyll to read them as plaintext instead, which is what we want so they display properly.

Here's what it's currently generating:
Screen Shot 2021-08-23 at 2 03 45 PM

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+```
+
+Note that in the above you'll need to replace `<NAMESPACE>` and `<RELEASE_NAME>` with the values
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:302]:
```suggestion
Replace `{NAMESPACE}` and `{RELEASE_NAME}` with the values
```

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+you originally provided to `helm install` in your own environment.
+
+Now that we've stored the IP in the environment variable `KONG_UDP_ENDPOINT` we can use
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:305]:
```suggestion
Now that you've stored the IP in the environment variable `KONG_UDP_ENDPOINT`, you can use
```

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+```shell
+$ dig @${KONG_UDP_ENDPOINT} konghq.com
+;; ANSWER SECTION:
+konghq.com.		30	IN	A	34.83.126.248
+
+;; Query time: 60 msec
+;; SERVER: <KONG_UDP_ENDPOINT>#9999
+;; WHEN: Thu Aug 19 08:39:16 EDT 2021
+;; MSG SIZE  rcvd: 77
+```
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:318]: Split the command from the response:

{% navtabs codeblock %}
{% navtab Command %}
```shell
dig @${KONG_UDP_ENDPOINT} konghq.com
```
{% endnavtab %}
{% navtab Response %}
```shell
;; ANSWER SECTION:
konghq.com.		30	IN	A	34.83.126.248
;; Query time: 60 msec
;; SERVER: <KONG_UDP_ENDPOINT>#9999
;; WHEN: Thu Aug 19 08:39:16 EDT 2021
;; MSG SIZE  rcvd: 77
```
{% endnavtab %}
{% endnavtabs %}

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+
+Verify that the `<KONG_UDP_ENDPOINT>` in the `SERVER` section of the response above ends
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:320]:
```suggestion
Verify that the `{KONG_UDP_ENDPOINT}` in the `SERVER` section of the response above ends
```

**Review by lena-larionova — CHANGES_REQUESTED** [review]
@shaneutt Jennifer is busy with Kuma and Mesh this week, so she asked me to review instead. 

I've left some phrasing, grammar, formatting edits (didn't test the functionality). Let me know if you have any questions, or if you can't get the tabs to work or something; I couldn't use Github suggestions for some things, as they get confused with fenced codeblocks inside the suggestion :( 

One more thing: can you add this file to the table of contents? You can find the data file under `app/_data/docs_nav_kic_2.0.x.yml`.

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+
+## Deploying CoreDNS
+
+For this example, we will deploy a default CoreDNS server that will do nothing more
+than serve up DNS requests.
+
+First we're going to need to configure CoreDNS with a [Core File][corefile] in order for
+it to properly serve DNS requests:
+
+```yaml
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:83]: Changed in 9d4f5af99201d9ff45b71705d220b595226c279f

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+Wait for the `LoadBalancer` service to be ready for the Gateway (see `kubectl get services`)
+and once it's up Kong is ready to serve UDP traffic on port `9999` external to the cluster
+using `UDPIngress` resources.
+
+## Deploying UDPIngress
+
+Now that Kong is listening on `9999` we can create a `UDPIngress` resource which will attach
+our CoreDNS service to that port so we can make DNS requests to it from outside the cluster.
+
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:274]: Changed in 9d4f5af99201d9ff45b71705d220b595226c279f

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+metadata:
+  name: udpingress-example
+```
+
+Or you can simply use the shorthand version with `kubectl`:
+
+```shell
+$ kubectl create namespace udpingress-example
+```
+
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md]: Changed in 9d4f5af99201d9ff45b71705d220b595226c279f

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+`/etc/resolv.conf` (by default this will point to the standard `kube-dns`
+service provided by the cluster).
+
+Save the this [ConfigMap][cfgmap] as `corefile.yaml` and apply it:
+
+```shell
+$ kubectl apply -f corefile.yaml
+```
+
+Once that's complete we're ready for the [Deployment][deployment] manifest:
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md]: Changed in 9d4f5af99201d9ff45b71705d220b595226c279f

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+[deployment]:https://kubernetes.io/docs/concepts/workloads/controllers/deployment/
+
+## Exposing CoreDNS via Service
+
+A Kubernetes [Service][svc] is one of the fundamental network abstraction layer
+that allows you to load balance traffic to pods in the cluster. `Services` are
+ultimately the DNS names that the Kong Gateway will be routing our UDP traffic to.
+
+Exposing a `Deployment` via a service is easy, here's the manifest which exposes
+the CoreDNS `Deployment` we made in the previous section:
```
[the remark this answers is not in the document]
Comment (shaneutt) [reply app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md]: Changed in 9d4f5af99201d9ff45b71705d220b595226c279f

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md (hunk 4 above)
Comment (shaneutt) [reply app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md]: Changed in 9d4f5af99201d9ff45b71705d220b595226c279f

**Comment by shaneutt** [comment]
@lena-larionova I _believe_ I've addressed all comments thank you for your suggestions. Ready for another read through.

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+Apply the following manifest to the cluster to create the namespace:
+
+```yaml
+apiVersion: v1
+kind: Namespace
+metadata:
+  name: udpingress-example
+```
+
+{% endnavtag %}
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:60]:
```suggestion
{% endnavtab %}
```
Typo here, probably mine.

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 12]
```diff
@@ -0,0 +1,351 @@
...
+> **Note**: This feature is compatible with:
+> * Kong Gateway versions FIXME and above.
+> * Kong Ingress Controller versions 2.0.0 and above.
+
+## Create a namespace
+
+First, create a namespace for testing DNS services and `UDPIngress`.
+
+{% navtabs %}
+{% navtabs Manifest %}
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:49]:
```suggestion
{% navtab Manifest %}
```
One more (reviewing mid-meeting is error-prone...)

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 13]
```diff
@@ -0,0 +1,351 @@
...
+## Overview
+
+Some of the most common UDP-based services available on the internet are
+[DNS servers][dns]. These servers are an important part of the infrastructure
+of the internet, and are also [present by default][kubedns] in Kubernetes
+clusters to allow [Pods][pods] within the cluster to look up other pods' IP
+addresses by name.
+
+For this example, you will:
+* Deploy your own [CoreDNS][coredns] server, which is the default
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:21]: This shortlink doesn't resolve to anything in the output. Was this supposed to be [kubedns], or is there another link missing?

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 14]
```diff
@@ -0,0 +1,351 @@
+---
+title: UDPIngress with Kong
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:2]:
```suggestion
title: UDPIngress with Kong Gateway
```

**Review by lena-larionova — COMMENTED** [review]
Looking good - one minor issue with a link, missed an instance of "Kong", and still need to add the topic to the navigation file so that it appears in the table of contents:

> Can you add this file to the table of contents? You can find the data file under app/_data/docs_nav_kic_2.0.x.yml.

Add an entry to the [guides section](https://github.com/Kong/docs.konghq.com/blob/udp-guide/app/_data/docs_nav_kic_2.0.x.yml#L52) in this format:
```
- title: UDPIngress with Kong Gateway
   url: /kubernetes-ingress-controller/guides/using-udpingress
```

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+[pods]:https://kubernetes.io/docs/concepts/workloads/pods/
+[chart]:https://github.com/kong/charts
+
+## Installation
+
+Follow the [deployment](/kubernetes-ingress-controller/{{page.kong_version}}/deployment/overview) documentation to install
+the {{site.kic_product_name}} on your Kubernetes cluster.
+
+> **Note**: This feature is compatible with:
+> * Kong Gateway versions FIXME and above.
```
Comment (lena-larionova) [inline app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:41]: Do we know which versions of the gateway this is compatible with yet, or does that still need to be tested?

**Comment by shaneutt** [comment]
> Looking good - one minor issue with a link, missed an instance of "Kong", and still need to add the topic to the navigation file so that it appears in the table of contents:
> 
> > Can you add this file to the table of contents? You can find the data file under app/_data/docs_nav_kic_2.0.x.yml.
> 
> Add an entry to the [guides section](https://github.com/Kong/docs.konghq.com/blob/udp-guide/app/_data/docs_nav_kic_2.0.x.yml#L52) in this format:
> 
> ```
> - text: UDPIngress with Kong Gateway
>   url: /guides/using-udpingress
> ```

cf6868fe9dafe377da09d0995816cf9b00fa6104 adds the nav :+1:

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md (hunk 13 above)
Comment (shaneutt) [reply app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md:21]: I missed the link, good catch thank you now fixed in 1513aa547490817ee8ea05712cda81025c216b91

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md (hunk 14 above)
Comment (shaneutt) [reply app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md]: :+1: fixed in 1513aa547490817ee8ea05712cda81025c216b91

File: app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md (hunk 15 above)
Comment (shaneutt) [reply app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress.md]: requires 2.0.0+, fix in 1513aa547490817ee8ea05712cda81025c216b91 :+1:

**Comment by shaneutt** [comment]
Should be ready again @lena-larionova :bow:

**Review by lena-larionova — APPROVED** [review]
Everything works! ✨  Checked the deploy preview at https://deploy-preview-3123--kongdocs.netlify.app/kubernetes-ingress-controller/2.0.x/guides/using-udpingress/

LGTM - @shaneutt , thanks for your patience.

**Pull request merged**